packages/react-reconciler/src/ReactFiberNewContext.new.js

```diff
@@ -311,6 +359,98 @@ export function propagateContextChange<T>(
   }
 }
 
+export function lazilyPropagateParentContextChanges(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+) {
+  propagateParentContextChanges(current, workInProgress, renderLanes, true);
+}
+
+export function propagateParentContextChangesToDeferredTree(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+) {
+  propagateParentContextChanges(current, workInProgress, renderLanes, false);
+}
+
+function propagateParentContextChanges(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+  forcePropagateEntireTree: boolean,
+) {
+  if (!enableLazyContextPropagation) {
+    return false;
+  }
+
+  let parent = workInProgress;
+  while (parent !== null && (parent.flags & DidPropagateContext) === NoFlags) {
+    if (parent.tag === ContextProvider) {
+      const currentParent = parent.alternate;
+      if (currentParent !== null) {
+        const oldProps = currentParent.memoizedProps;
+        if (oldProps !== null) {
+          const providerType: ReactProviderType<any> = parent.type;
+          const context: ReactContext<any> = providerType._context;
+
+          const newProps = parent.pendingProps;
+          const newValue = newProps.value;
+
+          const oldValue = oldProps.value;
+
+          const changedBits = calculateChangedBits(context, newValue, oldValue);
+          if (changedBits !== 0) {
+            // The context value changed. Search for matching consumers and
+            // schedule them to update.
+            propagateContextChangeUnlessAlreadyHasPendingWork(
+              workInProgress,
+              context,
+              changedBits,
+              renderLanes,
+              forcePropagateEntireTree,
+            );
+          }
+        }
+      }
+    }
+    parent = parent.return;
+  }
+
+  // This is an optimization so that we only propagate each provider once per
+  // subtree. (We will propagate the same provider to different subtrees, though
+  // — that's why the flag is on the fiber that bailed out, not the provider.)
+  // If a deeply nested child bails out, and it calls this propagation function,
+  // it uses this flag to know that the remaining ancestor providers have
+  // already been propagated.
+  workInProgress.flags |= DidPropagateContext;
```

Comment:
I need to think more about this part a little bit. I think this only works if you propagate completely without bailing out on paths that were marked by another context. Gotta think through that a bit more.

The way I had it before was it would only set this flag if none of the propagations bailed out. But I'm trying to avoid that because that can sometimes lead to propagating the same tree with the same provider multiple times.

Reply:
I'm pretty confident we can think of a clever trick here. For now I'll remove the `childLanes` thing and always propagate the whole subtree. Will tackle the many different providers problem in a subsequent commit.